[Security Solution] Editing rules independently of source data (#180407)

e40pud opened the pull request:
## Summary

Addresses https://github.com/elastic/kibana/issues/180407
Addresses https://github.com/elastic/kibana/issues/178611

With these changes we allow user to create and update a rule even if there are certain query bar validation error exist. Right now, we will make any non-syntax validation errors in `EQL` and `ES|QL` rules types to be non-blocking during the rule creation and rule updating workflows.

### Screenshot of the EQL rule creation workflow with existing non-blocking validation errors:

https://github.com/user-attachments/assets/06b7f76c-e600-4a99-8ead-1445d429e9d3

### Screenshot of the EQL rule updating workflow with existing non-blocking validation errors:

https://github.com/user-attachments/assets/9b35e113-b127-487b-bc23-afecf704db9d

## UPDATE

After discussing confirmation modal with @approksiu, we decided to simplify it and show only title with generic description to avoid too be too literal in the modal. User can see the full error description during rule creation/editing workflows in the query bar where we show each validation error as part of the query bar form item.

<img width="702" alt="Screenshot 2024-08-28 at 12 50 14" src="https://github.com/user-attachments/assets/edfb791e-4e45-4fa5-8a46-c7e2772abdf9">

### Some test cases for local testing

<details>
  <summary><b>Create EQL rule with missing data source</b></summary>

#### Steps:
1. Open rules management page
2. Click create new rule button
3. Select EQL rule type
4. Set non-existing index in index patterns field
5. Add some valid EQL query (for example `any where true`)
6. Continue with other steps
7. Click create rule button

**Expected**: You will see the confirmation modal that warns user about potentially failing rule executions. Clicking `Confirm` button will create a rule.

</details>

<details>
  <summary><b>Create EQL rule with missing data field</b></summary>

#### Steps:
1. Open rules management page
2. Click create new rule button
3. Select EQL rule type
4. Set existing indices in index patterns field
5. Add some valid EQL query referring non-existing data field (for example `any where agent.non_existing_field`)
6. Continue with other steps
7. Click create rule button

**Expected**: You will see the confirmation modal that warns user about potentially failing rule executions. Clicking `Confirm` button will create a rule.

</details>

<details>
  <summary><b>Create EQL rule with syntax error in the query</b></summary>

#### Steps:
1. Open rules management page
2. Click create new rule button
3. Select EQL rule type
4. Set existing indices in index patterns field
5. Add some syntactically invalid EQL query (for example `hello world`)

**Expected**: The continue button does not allow user to proceed to the About step due to existing syntax error.

</details>

<details>
  <summary><b>Create ES|QL rule with missing data source</b></summary>

#### Steps:
1. Open rules management page
2. Click create new rule button
3. Select ES|QL rule type
4. Add some valid ES|QL query with non-existing data source (for example `from non-existing-index-* metadata _id, _version, _index | SORT @timestamp`)
6. Continue with other steps
7. Click create rule button

**Expected**: You will see the confirmation modal that warns user about potentially failing rule executions. Clicking `Confirm` button will create a rule.

</details>

<details>
  <summary><b>Create ES|QL rule with missing data field</b></summary>

#### Steps:
1. Open rules management page
2. Click create new rule button
3. Select ES|QL rule type
4. Add some valid ES|QL query with non-existing data field (for example `from logs-* metadata _id, _version, _index | SORT agent.non_existing_field`)
6. Continue with other steps
7. Click create rule button

**Expected**: You will see the confirmation modal that warns user about potentially failing rule executions. Clicking `Confirm` button will create a rule.

</details>

<details>
  <summary><b>Create ES|QL rule with syntax error in the query</b></summary>

#### Steps:
1. Open rules management page
2. Click create new rule button
3. Select ES|QL rule type
4. Add some syntactically invalid ES|QL query (for example `hello world`)

**Expected**: The continue button does not allow user to proceed to the About step due to existing syntax error.

</details>

Same behaviour applies to the rule updating workflow. For example, you can try to install one of the EQL or ES|QL rules that point to non-existing data source or uses non-existing data field. User can still update (add rule actions) to such installed pre-built rules.

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
  * https://github.com/elastic/security-docs/issues/5758
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
  * [Detection Engine - Cypress](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6831) (100 ESS & 100 Serverless)
  * [Rule Management - Cypress](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6830) (100 ESS & 100 Serverless)

e40pud committed:
[Security Solution] Editing rules independently of source data (#180407)

e40pud commented:
/ci

e40pud committed:
Fix tests

e40pud commented:
/ci

elasticmachine committed:
Merge branch 'main' into security/feature/180407-save-rules

e40pud commented:
/ci

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

e40pud commented:
/ci

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-engine (Team:Detection Engine)

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6817](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6817)
[✅] Security Solution Rule Management - Cypress: 100/100 tests passed.
[❌] [Serverless] Security Solution Rule Management - Cypress: 75/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F191487%2Fhead)

e40pud committed:
Simplify confirmation modal description

e40pud commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'main' into security/feature/180407-save-rules

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6816](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6816)
[❌] Security Solution Detection Engine - Cypress: 49/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 33/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F191487%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6827](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6827)
[❌] Security Solution Detection Engine - Cypress: 0/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 0/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F191487%2Fhead)

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/pages/rule_editing/index.tsx:
```diff
@@ -398,43 +440,39 @@ const EditRulePageComponent: FC<{ rule: RuleResponse }> = ({ rule }) => {
+    saveChanges,
   ]);
 
+  const showSaveWithErrorsModal = () => setIsSaveWithErrorsModalVisible(true);
+  const closeSaveWithErrorsModal = () => setIsSaveWithErrorsModalVisible(false);
```
Component `SaveWithErrorsModal` (x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/save_with_errors_confirmation/index.tsx) is wrapped in memo, but these callbacks will be recreated on every re-render. So, memo won't work as intended

vitaliidm replied:
same for `x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/pages/rule_creation/index.tsx` file as well

e40pud replied:
good catch!

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/pages/form.tsx:
```diff
@@ -156,3 +160,84 @@ export const useRuleIndexPattern = ({
+  if (fieldError.message.length > 0) {
+    return [fieldError.message];
+  } else if (Array.isArray(fieldError.messages)) {
+    return fieldError.messages.map((message) => JSON.stringify(message));
```
why message is stringified here but not at 173 line?

Maybe simple unit tests or comment can be helpful

e40pud replied:
Added comment

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/pages/rule_creation/index.tsx:
```diff
@@ -559,7 +633,8 @@ const CreateRulePageComponent: React.FC = () => {
     () =>
-      defineStepForm.isValid && (
+      defineStepForm.isValid !== undefined &&
+      activeStep !== RuleStep.defineRule && (
```
why extra condition is needed here?

e40pud replied:
added comments

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/save_with_errors_confirmation/index.tsx:
```diff
@@ -0,0 +1,41 @@
+ * 2.0.
+ */
+
+import React from 'react';
```
@e40pud 

is this fine user that user sees some error, but can't go to Definition tab or see error itself?

<img width="2375" alt="Screenshot 2024-08-29 at 10 24 40" src="https://github.com/user-attachments/assets/d0a03e4c-82e2-473d-99c0-a80f3d196a48">

I imagine it will be a case until full support of prebuilt rule editing will be released.
But before that, it can affect Serverless releases

e40pud replied:
What you see right now is the current behaviour on both ESS and Serverless. Users are unable to modify prebuilt rules. The whole ["rule customization"](https://github.com/elastic/security-team/issues/8056) effort will allow user to do that.

vitaliidm replied:
I am aware, prebuilt rule can't be modified. 
It's the fact that we show user step is invalid without showing actual error and additional information how that can affect rule's behavior.

Is there a way for user in this situation to see actual error?

e40pud replied:
As discussed here is the ticket to discuss this behaviour https://github.com/elastic/kibana/issues/191832

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/pages/form.tsx:
```diff
@@ -156,3 +160,84 @@ export const useRuleIndexPattern = ({
+};
+
+export const useRuleFormsErrors = () => {
+  const getRuleFormsErrors = useCallback(
```
would be great to have unit tests for this function/hook if possible

e40pud replied:
added some basic tests

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/pages/rule_creation/index.tsx:
```diff
@@ -328,104 +335,177 @@ const CreateRulePageComponent: React.FC = () => {
+      const { valid, blockingErrors, nonBlockingErrors } = await validateEachStep();
+      if (valid) {
+        // There are no validation errors, thus proceed to rule creation
+        createRuleFromFormData(enabled);
```
this function is asynchronous. should we wait until it resolves?

e40pud replied:
done

vitaliidm reviewed (commented):
thanks for implementing this @e40pud 
Looks good, just few minor comments from me

e40pud committed:
Update x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/pages/translations.ts

Co-authored-by: Vitalii Dmyterko <92328789+vitaliidm@users.noreply.github.com>

e40pud commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'main' into security/feature/180407-save-rules

e40pud committed:
Review feedback

- memoize callbacks
- clarify multiple EQL validation errors functionality
- comments to clarify step editing button visibility conditions
- add unit tests for `useRuleFormsErrors`

e40pud commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'main' into security/feature/180407-save-rules

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6830](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6830)
[❌] Security Solution Rule Management - Cypress: 75/100 tests passed.
[✅] [Serverless] Security Solution Rule Management - Cypress: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F191487%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6828](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6828)
[❌] Security Solution Detection Engine - Cypress: 30/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 42/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F191487%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6831](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6831)
[❌] Security Solution Detection Engine - Cypress: 49/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 36/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F191487%2Fhead)
